### ISSUE TITLE
regression when upgrading from 4.1.1 to 4.1.2

### ISSUE THREAD

**Issue by amit777** [Automattic/mongoose#3253]
My test suite passes with 4.1.1 and before when trying to get a stream from GridStore. After updating to 4.1.2, my test breaks.  Here is the code that returns a stream:

```
var mongoose = require('mongoose');
var GridStore = mongoose.mongo.GridStore;
attachmentSchema.methods.getStream = function(){
  var that = this;
  return co(function*(){
    var db = conn.db;
    var gfs = new GridStore(db, `/opt/data/${that.hash}`, "r", {root: 'fs'} );
    return gfs.stream();
  });
};
```

It seems to affect the conn.db object.  It's not clear to me if this is the real reason for my issue or not.  Basically my mocha test returns an error quickly with 4.1.2.  In 4.1.1, my mocha test seems to wait a little while for the gridfs db connection or mongo connection to establish. Then I get my file stream.

The weird thing is my code seems to be runnign find in production.  My test suite may be running into a race condition where the db connection is not established.

Apologies if my explanation is not clear.

**Comment by amit777** [comment]
if i use 4.1.2 and downgrade the mongo native driver dependency to 2.0.34, then my tests work again.  So it seems like my issue is caused by mongodb native driver update..

**Comment by amit777** [comment]
and to be more specific, it seems like my issue occurs between  2.0.39 and 2.0.40 on the native driver. I guess I will check with that project team.

**Comment by vkarpov15** [comment]
What does the test failure look like?

**Comment by amit777** [comment]
I have an issue in my mocha tests getting a stream.. Seems to work fine when my app is running though.. Here is some code around the issue:

```
var conn;
var connect = function(){
   console.log(`connecting to gridfs:${config.get('gridfs.db')}`);
   conn = mongoose.createConnection(config.get('gridfs.db'), config.get('gridfs.dbOptions'));
   conn.on('error', console.log);
   conn.on('disconnected', connect);
   conn.on('connected', function(){
     console.log(`DEBUGGED CONNECTED ATTACHMENT`);
   });
};
connect();

attachmentSchema.methods.getStream = function(){
  var that = this;
  return co(function*(){
    var db = conn.db;
    var gfs = new GridStore(db, `/opt/sendergen/data/${that.hash}`, "r", {root: 'fs'} );
    return gfs.stream();
  });
};

```

**Comment by amit777** [comment]
ok.. i've narrowed it down to this commit:
https://github.com/christkv/mongodb-core/commit/f600b7449bcf2707a9b84fd1c5c14f5236019941

i'm using a replicaset.. when I revert this file, my test works.  The test takes a little time to complete and connect to the replicaset.. but it succeeds.. when I apply the above commit, my test comes back immediately with an error.  I'm sorry I'm a bit new to all this.  It's not clear to me why, but it seems like some sort of race condition.

**Comment by vkarpov15** [comment]
Can you show me what the error message is? Hard to tell with just the commit what's going wrong.

**Comment by amit777** [comment]
Hi, sorry I forgot I left this ticket open. It turned out to be an issue that I caused. It basically happened because I had 2 different modules that were connecting to mongoose in the same app.  I think some race condition was created in the native driver.  .  The issue was fixed when I removed te second connection to mongoose.  Here is a link to some more info if it's helpful:
https://github.com/mongodb/node-mongodb-native/pull/1289#issuecomment-132320700
